Feature engine

rk4514 opened the pull request:
This is the mostly completed game engine. Only thing left to do is implement trading and communication with sockets. I've included some functions that interface with the console for the sake of testing but those will be removed later.

rk4514 committed:
Implement more game functionality

-playing of development cards
-longest road calculations.
-largest army.
-victory points and winning conditions.

rk4514 committed:
Add harbors

jackadrianglass commented on batan-server/engine/board.ts:
```diff
@@ -158,11 +160,34 @@ export class Board {
+    }
+
+    private initHarbors() {
+        this.vertexList[0].harbor = harborType.threeForOne;
+        this.vertexList[3].harbor = harborType.threeForOne;
+        this.vertexList[1].harbor = harborType.wheat;
+        this.vertexList[5].harbor = harborType.wheat;
+        this.vertexList[10].harbor = harborType.ore;
+        this.vertexList[15].harbor = harborType.ore;
+        this.vertexList[26].harbor = harborType.threeForOne;
+        this.vertexList[32].harbor = harborType.threeForOne;
+        this.vertexList[42].harbor = harborType.sheep;
+        this.vertexList[46].harbor = harborType.sheep;
+        this.vertexList[49].harbor = harborType.threeForOne;
+        this.vertexList[52].harbor = harborType.threeForOne;
+        this.vertexList[51].harbor = harborType.threeForOne;
+        this.vertexList[47].harbor = harborType.threeForOne;
+        this.vertexList[38].harbor = harborType.brick;
+        this.vertexList[33].harbor = harborType.brick;
+        this.vertexList[16].harbor = harborType.wood;
+        this.vertexList[11].harbor = harborType.wood;
```
Could this be put into a map and iterated through? Like {idx: type}

rk4514 replied:
Yes but for sake of time i'm going to leave this as is.

jackadrianglass commented on batan-server/engine/board.ts:
```diff
@@ -242,97 +241,97 @@ export class Board {
-            this.hexList[i].vertices = [0+1, 3+i, 4+i, 7+i, 8+i, 12+i];
+    private setHexVertices() {
+        for (let i = 0, j = 0; i < 3; i++, j++) {
+            this.hexList[i].vertices = [0 + j, 3 + j, 4 + j, 7 + j, 8 + j, 12 + j];
```
couldn't you use i instead of j? Their the same value the entire time

jackadrianglass reviewed (approved):
Looks good to me. I had a couple of nits

lhossackucalgary commented on batan-server/engine/game.ts:
```diff
@@ -99,93 +194,494 @@ class Game {
+      this.moveRobberAndSteal(user);
+
+      //discard if more than 7
+      this.players.forEach((player) => {
```
Robber should be moved _after_ players discard half their cards if they have over 7

rk4514 replied:
Done, nice catch!

lhossackucalgary commented on batan-server/engine/game.ts:
```diff
@@ -99,93 +194,494 @@ class Game {
+      this.players.forEach((player) => {
+        let totalResources = player.resources.brick + player.resources.ore + player.resources.sheep + player.resources.wheat + player.resources.wood;
+        if (totalResources >= 7) {
+          //TODO emit to user... this will be handled with sockets later ... need to ensure correct number discarded
```
I think we'll probably need to split this into a few functions or include an additional state machine for context of being part way through this operation during the back & forth between client/server

rk4514 replied:
@lhossackucalgary for sake of time, we can just make it take the appropriate number of cards at random if that is easier. Should not affect game play all that much.

lhossackucalgary reviewed (approved):
lgtm for this merge. 
For the future, I have some concerns about some of the operations like when a 7 is rolled (& robber is moved) that require user input part way through the operation. I think it can be resolved by adding some additional state that makes other moves invalid until everyone with over 7 cards has tossed half their cards for example, but it definitely throws a bit of a wrench into the strictly turn-based flow (& trading will as well). I don't know if we will even get to implementing trading on the front end with the 10 days we have remaining though

rk4514 committed:
Address feedback

-extract diceRoll to return value to front end.
-remove unused iterator.
-reorder events when 7 is rolled.
-make resources discard automatically at random.
-minor documentation fixes.

rk4514 committed:
Add trade with bank functionality

-formatting.

rkettles14 commented:
@jackadrianglass @lhossackucalgary Thanks for the feedback! Comments addressed.
